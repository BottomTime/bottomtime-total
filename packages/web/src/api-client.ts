import { ApiClient } from '@bottomtime/api';

import { InjectionKey, inject } from 'vue';

export const ApiClientKey: InjectionKey<ApiClient> = Symbol('ApiClient');
export function useClient(): ApiClient {
  const client = inject(ApiClientKey);

  if (!client) {
    throw new Error(
      'Api Client has not been initialized. Did you forget to call `app.provide(key, client)`?',
    );
  }

  return client;
}
