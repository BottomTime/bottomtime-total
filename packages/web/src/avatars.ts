import { AvatarSize, ListAvatarURLsResponseDTO } from '@bottomtime/api';

export type AvatarURLs = ListAvatarURLsResponseDTO['sizes'];

export function getAvatarURL(
  baseUrl: string | undefined,
  size: AvatarSize,
): string | undefined {
  if (!baseUrl) return undefined;

  if (/^\/api\/users\/[a-z0-9]+([_.-][a-z0-9]+)*\/avatar/i.test(baseUrl)) {
    return baseUrl.endsWith('/') ? `${baseUrl}${size}` : `${baseUrl}/${size}`;
  }

  return baseUrl;
}
