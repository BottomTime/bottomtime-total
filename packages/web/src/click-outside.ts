import { Directive, DirectiveBinding } from 'vue';

type HTMLElementWithClickOutside = HTMLElement & {
  clickOutsideEvent: (e: MouseEvent) => void;
};

export const clickOutside: Directive = {
  beforeMount: (el: HTMLElementWithClickOutside, binding: DirectiveBinding) => {
    el.clickOutsideEvent = (event) => {
      // here I check that click was outside the el and his children
      if (!(el == event.target || el.contains(event.target as Node))) {
        // and if it did, call method provided in attribute value
        binding.value();
      }
    };
    document.addEventListener('click', el.clickOutsideEvent);
  },
  unmounted: (el: HTMLElementWithClickOutside) => {
    document.removeEventListener('click', el.clickOutsideEvent);
  },
};
