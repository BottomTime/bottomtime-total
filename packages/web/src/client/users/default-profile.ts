import { assertValid } from '@/helpers';
import { SuperAgentStatic } from 'superagent';
import {
  Profile,
  ProfileCertificationData,
  ProfileData,
  ProfileDataSchema,
  UserData,
} from './interfaces';
import { ProfileVisibility } from '@/constants';

export class DefaultProfile implements Profile {
  private static readonly JsonOmitKeys = new Set([
    'userId',
    'memberSince',
    'username',
  ]);

  constructor(
    private readonly agent: SuperAgentStatic,
    private readonly data: UserData,
  ) {}

  get userId(): string {
    return this.data.id;
  }
  get memberSince(): Date {
    return this.data.memberSince;
  }
  get username(): string {
    return this.data.username;
  }

  get avatar(): string | undefined {
    return this.data.profile.avatar;
  }
  set avatar(value: string | undefined) {
    this.data.profile.avatar = value;
  }

  get bio(): string | undefined {
    return this.data.profile.bio;
  }
  set bio(value: string | undefined) {
    this.data.profile.bio = value;
  }

  get birthdate(): string | undefined {
    return this.data.profile.birthdate;
  }
  set birthdate(value: string | undefined) {
    this.data.profile.birthdate = value;
  }

  get customData(): Record<string, unknown> | undefined {
    return this.data.profile.customData;
  }
  set customData(value: Record<string, unknown> | undefined) {
    this.data.profile.customData = value;
  }

  get certifications(): ProfileCertificationData[] | undefined {
    return this.data.profile.certifications;
  }
  set certifications(value: ProfileCertificationData[] | undefined) {
    this.data.profile.certifications = value;
  }

  get experienceLevel(): string | undefined {
    return this.data.profile.experienceLevel;
  }
  set experienceLevel(value: string | undefined) {
    this.data.profile.experienceLevel = value;
  }

  get location(): string | undefined {
    return this.data.profile.location;
  }
  set location(value: string | undefined) {
    this.data.profile.location = value;
  }

  get name(): string | undefined {
    return this.data.profile.name;
  }
  set name(value: string | undefined) {
    this.data.profile.name = value;
  }

  get profileVisibility(): ProfileVisibility {
    return this.data.profile.profileVisibility;
  }
  set profileVisibility(value: ProfileVisibility) {
    this.data.profile.profileVisibility = value;
  }

  get startedDiving(): string | undefined {
    return this.data.profile.startedDiving;
  }
  set startedDiving(value: string | undefined) {
    this.data.profile.startedDiving = value;
  }

  async save(): Promise<void> {
    const profile = assertValid<ProfileData>(
      this.data.profile,
      ProfileDataSchema,
    );
    await this.agent.put(`/api/profiles/${this.username}`).send(profile);
    this.data.profile = profile;
  }

  toJSON(): ProfileData {
    return assertValid(this.data.profile, ProfileDataSchema);
  }
}
