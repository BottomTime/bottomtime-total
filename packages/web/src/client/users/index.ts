export * from './interfaces';

export * from './default-user';
export * from './default-user-manager';
