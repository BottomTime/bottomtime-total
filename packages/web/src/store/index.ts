export * from './alerts.store';
export * from './current-user.store';
export * from './operators.store';
export * from './dive-sites.store';
export * from './error.store';
export * from './friends.store';
export * from './log-entries.store';
export * from './profiles.store';
export * from './tanks.store';
export * from './toasts.store';
