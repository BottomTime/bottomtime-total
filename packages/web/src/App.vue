<template>
  <div v-if="isLoading"></div>
  <div v-else>
    <SnackBar />
    <NavBar />
    <RouterView />
    <PageFooter />
    <CookieWarning />
  </div>
</template>

<script lang="ts" setup>
import { onBeforeMount, ref } from 'vue';

import CookieWarning from './components/main/CookieWarning.vue';
import { Dispatch, useStore } from '@/store';
import { inject } from '@/helpers';
import NavBar from '@/components/main/NavBar.vue';
import PageFooter from '@/components/main/PageFooter.vue';
import SnackBar from '@/components/main/SnackBar.vue';
import { ApiClientKey } from '@/injection-keys';

const client = inject(ApiClientKey);
const store = useStore();
const isLoading = ref(true);

onBeforeMount(async () => {
  try {
    const currentUser = await client.users.getCurrentUser();
    await store.dispatch(Dispatch.SignInUser, currentUser);
  } catch (error) {
    console.error(error);
  } finally {
    isLoading.value = false;
  }
});
</script>

<style>
@import './assets/bulmaswatch.min.css';
</style>
