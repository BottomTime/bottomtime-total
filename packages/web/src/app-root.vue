<template>
  <SnackBar />
  <div
    id="modal"
    class="font-content text-grey-900 bg-blue-200 dark:text-grey-100 dark:bg-blue-950"
  ></div>
  <div id="drawer"></div>

  <div
    class="md:container mx-auto font-content text-grey-900 bg-blue-200 dark:text-grey-100 dark:bg-blue-950 rounded-b-xl shadow-md shadow-white dark:shadow-grey-700 opacity-95"
  >
    <!-- Placeholder element. Modal dialogs will be "teleported" here to get the z-indexing right. -->

    <section class="mt-16 p-4">
      <RouterView></RouterView>
    </section>
    <PageFooter />
  </div>
  <NavBar />
</template>

<script setup lang="ts">
import { onBeforeMount } from 'vue';

import { ToastType } from './common';
import NavBar from './components/core/nav-bar.vue';
import PageFooter from './components/core/page-footer.vue';
import SnackBar from './components/core/snack-bar.vue';
import { Config } from './config';
import { useErrors, useToasts } from './store';

const toasts = useToasts();
const errors = useErrors();

onBeforeMount(() => {
  if (!Config.isSSR) {
    if (errors.renderError) {
      if (!Config.isProduction) {
        /* eslint-disable-next-line no-console */
        console.error(errors.renderError);
      }
      toasts.toast({
        id: 'server-prerender-error',
        message:
          'A server error has occurred and the page may not have loaded correctly. We apologize for the inconvenience. Please try again later as we investigate the issue.',
        type: ToastType.Error,
      });
    }
  }
});
</script>
