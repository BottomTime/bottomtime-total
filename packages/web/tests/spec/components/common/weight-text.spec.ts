import { UserDTO, WeightUnit } from '@bottomtime/api';

import { ComponentMountingOptions, mount } from '@vue/test-utils';

import { Pinia, createPinia } from 'pinia';

import WeightText from '../../../../src/components/common/weight-text.vue';
import { useCurrentUser } from '../../../../src/store';
import { BasicUser } from '../../../fixtures/users';

describe('WeightText component', () => {
  [
    { name: 'kilograms', unit: WeightUnit.Kilograms, weight: 2.25 },
    { name: 'pounds', unit: WeightUnit.Pounds, weight: 5.5 },
  ].forEach(({ name, unit, weight }) => {
    let pinia: Pinia;
    let currentUser: ReturnType<typeof useCurrentUser>;
    let opts: ComponentMountingOptions<typeof WeightText>;

    beforeEach(() => {
      pinia = createPinia();
      currentUser = useCurrentUser(pinia);
      opts = {
        props: {
          weight,
          unit,
        },
        global: {
          plugins: [pinia],
        },
      };
    });

    it(`will display in natural units (${name}) for unauthenticated users`, () => {
      currentUser.user = null;
      const wrapper = mount(WeightText, opts);
      expect(wrapper.text()).toMatchSnapshot();
    });

    it(`will display in the user's preferred units (${name}) without conversion if possible`, () => {
      const user: UserDTO = {
        ...BasicUser,
        settings: {
          ...BasicUser.settings,
          weightUnit: unit,
        },
      };
      currentUser.user = user;
      const wrapper = mount(WeightText, opts);
      expect(wrapper.text()).toMatchSnapshot();
    });

    it(`will display in the user's preferred units with conversion from ${name} if necessary`, () => {
      const user: UserDTO = {
        ...BasicUser,
        settings: {
          ...BasicUser.settings,
          weightUnit:
            unit === WeightUnit.Kilograms
              ? WeightUnit.Pounds
              : WeightUnit.Kilograms,
        },
      };
      currentUser.user = user;
      const wrapper = mount(WeightText, opts);
      expect(wrapper.text()).toMatchSnapshot();
    });
  });
});
