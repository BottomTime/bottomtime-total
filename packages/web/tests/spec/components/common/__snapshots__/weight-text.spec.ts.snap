// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WeightText component will display in natural units (kilograms) for unauthenticated users 1`] = `"2.3 kg"`;

exports[`WeightText component will display in natural units (pounds) for unauthenticated users 1`] = `"5.5 lb"`;

exports[`WeightText component will display in the user's preferred units (kilograms) without conversion if possible 1`] = `"2.3 kg"`;

exports[`WeightText component will display in the user's preferred units (pounds) without conversion if possible 1`] = `"5.5 lb"`;

exports[`WeightText component will display in the user's preferred units with conversion from kilograms if necessary 1`] = `"5.0 lb"`;

exports[`WeightText component will display in the user's preferred units with conversion from pounds if necessary 1`] = `"2.5 kg"`;
